ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala

```diff
@@ -455,6 +473,36 @@ private[kvutils] class TransactionCommitter(
           RejectionReason.Inconsistent("Causal monotonicity violated")))
   }
 
+  /**
+    * Validates that 'contract key' -> 'contract id' mapping resolved by a participant is still valid.
+    */
+  private def validateContractKeysConsistency(
+      recordTime: Option[Timestamp],
+      transactionEntry: DamlTransactionEntrySummary,
+      resolvedKeys: Map[DamlContractKey, String]
+  )(): StepResult[DamlTransactionEntrySummary] = {
+    val keysAreConsistent = transactionEntry.contractKeyToIdMap.forall {
+      case (contractKey, submittedContractId) =>
+        resolvedKeys.get(contractKey) match {
+          case None =>
+            // This means that there is no such key yet in the context.
+            // We allow such submission because the submitted contractKey is ~most likely~ a result of a create command,
+            // however a more strict validation should be done here to verify that the command type was really create.
+            true
+          case contractId: Some[String] =>
+            contractId == submittedContractId
+        }
+    }
+    if (keysAreConsistent) StepContinue(transactionEntry)
+    else
+      reject(
+        recordTime,
+        buildRejectionLogEntry(
+          transactionEntry,
+          RejectionReason.Inconsistent("Contract keys inconsistent")
```

Comment:
I don't know your policy on error message, but I'd include the inconsistent contract keys. This may be useful for debugging our own code as well as application code.

Reply:
I've added information about contract keys that not pass the validation.